mongosql/src/algebrizer/definitions.rs

```diff
@@ -904,32 +953,89 @@ impl<'a> Algebrizer<'a> {
+                // error.
+                ast::SelectBody::Standard(exprs) => match exprs.as_slice() {
+                    [ast::SelectExpression::Star] => {
+                        source.schema(&self.schema_inference_state())?;
+                        Ok(source)
+                    }
+                    _ => Err(Error::NonStarStandardSelectBody),
+                },
+                // SELECT VALUES expressions must be Substar expressions or normal Expressions that are
+                // Documents, i.e., that have Schema that Must satisfy ANY_DOCUMENT.
+                //
+                // All normal Expressions will be mapped as Datasource Bottom, and all Substars will be mapped
+                // as their name as a Datasource.
+                ast::SelectBody::Values(exprs) => {
+                    self.algebrize_select_values_body(exprs, source, is_add_fields, false)
+                }
+            },
+            ast::SetQuantifier::Distinct => match ast_node.body {
+                ast::SelectBody::Standard(exprs) => match exprs.as_slice() {
+                    [ast::SelectExpression::Star] => self.build_distinct_star_project_stage(source),
+                    _ => Err(Error::NonStarStandardSelectBody),
+                },
+                ast::SelectBody::Values(exprs) => {
+                    self.algebrize_select_values_body(exprs, source, is_add_fields, true)
                 }
-                _ => Err(Error::NonStarStandardSelectBody),
             },
-            // SELECT VALUES expressions must be Substar expressions or normal Expressions that are
-            // Documents, i.e., that have Schema that Must satisfy ANY_DOCUMENT.
-            //
-            // All normal Expressions will be mapped as Datasource Bottom, and all Substars will be mapped
-            // as their name as a Datasource.
-            ast::SelectBody::Values(exprs) => {
-                self.algebrize_select_values_body(exprs, source, is_add_fields)
-            }
         }
     }
 
+    fn build_distinct_star_project_stage(&self, source: mir::Stage) -> Result<mir::Stage> {
+        source.schema(&self.schema_inference_state())?;
+
+        let source_result_set = source.schema(&self.schema_inference_state())?;
+        let mut ds_names = source_result_set
+            .schema_env
+            .keys()
+            .filter_map(|key| match &key.datasource {
+                DatasourceName::Named(n) => Some((n.clone(), key.clone())),
+                DatasourceName::Bottom => None,
+            })
+            .collect::<Vec<(String, Key)>>();
+
+        let (table_name, _) = ds_names.remove(0);
+        let group_stage = mir::Stage::Group(mir::Group {
+            source: Box::new(source),
+            keys: vec![OptionallyAliasedExpr::Aliased(AliasedExpr {
+                alias: table_name.clone(),
+                expr: Expression::Reference(ReferenceExpr {
+                    key: Key::named(table_name.as_str(), self.scope_level),
```

Comment:
Shouldn't we be using the `key` from line 997 here?

Reply:
I updated this part a bit, I think it's now using the right key.